### ISSUE TITLE
qubes ctrl-shift-c/v collides with terminal c&p

### ISSUE THREAD

**Issue by ThomasWaldmann** [QubesOS/qubes-issues#4573]
Qubes 4.0, maybe also older.

There was already #3583 discussing that the qubes c&p key combo is the same as many people also use for terminal c&p.

I'ld like to suggest 2 changes to finally resolve it and avoid having people stumble over it again and again:

1. @marmarek mentioned "Try Ctrl+Ins and Shift+Ins. Works for me." - it also works for me, so could go into FAQ?

2. About `/etc/qubes/guid.conf`:

  - as people end up there, wanting to change keys, mention `Ctrl-Ins` / `Shift-Ins` there also, so maybe they don't even need to change after reading it there.

  - if there is really some need to change the keys, try and error there is rather tedious. as i did not know how to make changes in there apply without rebooting, i did reboot rather often until i found `Mod4-c` is the config value needed for "Windows-Key + c". xev names that key Super_L so it's not helpful here.

  - so, mention the valid names of the modifier keys there?

**Comment by ThomasWaldmann** [comment]
key names: "Ctrl", "Shift", "Mod4" (Windows key), ... (add more verified key names)

**Comment by ThomasWaldmann** [comment]
Proposed text for the config:
```
# Before changing secure_(copy|paste)_sequence below because you need
# Ctrl-Shift-(c|v) for normal terminal copy&paste, please note that you
# can also use Ctrl-Ins and Shift-Ins for that (thus: no need to change).
# If you still want to change, here is a list of valid modifier key names:
# "Ctrl", "Shift", "Mod4" (Windows key) 
```

**Comment by andrewdavidwong** [comment]
The documentation is a community effort. Please help us improve it by submitting a pull request:

https://www.qubes-os.org/doc/doc-guidelines/

**Comment by ThomasWaldmann** [comment]
@andrewdavidwong you have that on a F-key or so? :)

I'll do, just wanted to discuss it here first to get some feedback.

**Comment by airelemental** [comment]
I've always wondered, why is the default ctrl-shift-c? Why default to something that conflicts with terminal copy?

Wouldn't Super-v, Super-c be better? - It's memorable, it requires 2 keystrokes instead of 3, keys are close together (as opposed to Ctrl-Ins) and close to Ctrl-v/Ctrl-c (which you usually press soon after), and no conflict with keyboard shortcuts (usually Ctrl/Shift).

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong you have that on a F-key or so? :)

I have it as a "saved reply," because everyone suggests and requests documentation changes, while very few people actually submit them. Maybe I should say, "Be the change you want to see in the documentation." :slightly_smiling_face:

**Comment by h01ger** [comment]
hi, 

what terminal are you using that it conflicts with ctrl-shift-c?


-- 
cheers,
	Holger

-------------------------------------------------------------------------------
               holger@(debian|reproducible-builds|layer-acht).org
       PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C

**Comment by unman** [comment]
Both Konsole and gnome-terminal use Ctrl-Shift-(c|v) .

As to changing the default in Qubes, I guess any change will conflict with *someone*.
I wouldn't favour making the Qubes combo easier though, so @airelemental proposal doesn't work for me.

**Comment by andrewdavidwong** [comment]
We've already stated that we won't be changing the default keyboard shortcuts. Users are always free to [configure them](https://www.qubes-os.org/doc/config-files/) on their own systems. The defaults are so firmly established in Qubes by this point that the costs of changing them would outweigh any potential benefits.

**Comment by marmot1791** [comment]
This pull request addresses the comment by @ThomasWaldmann 
https://github.com/QubesOS/qubes-gui-daemon/pull/33

Additional documentation elsewhere may still be appropriate

**Comment by ddevz** [comment]
Note that even though Ctrl+Ins and Shift+Ins works, the edit menu of gnome-terminal (the default) still shows the user "Shift+Ctrl+C" and "Shift+Ctrl+V" as being the keybinding.  A new user should still be expected to believe what the menu is telling it.

Also note that I tried using edit->properties->shortcuts and changing the keys to "Ctrl+Insert" and "Shift+Insert" *does* display the correct keybinding then in the terminal edit menu, meaning this should be possible to fix by changing the default keybindings for the terminal.

**Comment by andrewdavidwong** [comment]
> Note that even though Ctrl+Ins and Shift+Ins works, the edit menu of gnome-terminal (the default) still shows the user "Shift+Ctrl+C" and "Shift+Ctrl+V" as being the keybinding.  A new user should still be expected to believe what the menu is telling it.
> 

Sounds like a gnome-terminal bug.

**Comment by ddevz** [comment]
> Sounds like a gnome-terminal bug.

If you mean bug in the general sense that it's specific to gnome-terminal configuration then yes.  In case this matters (and in case you don't already know this):  the "official" keybindings are "Shift-Ctrl-C".  This is what is in the default configuration on the "shortcuts" configuration screen.  the Ctrl-Insert keybinding seems to be a "secret" *second* keybinding that does not show up on the shortcuts configuration screen.  (i say "secret" cause it does not even appear on the gnome-terminal copy paste documentation at https://help.gnome.org/users/gnome-terminal/stable/txt-copy-paste.html )

Anyway, I was just proposing that if gnome-terminal is the default qubes terminal, and the default gnome-terminal keybinding of Shift-Ctrl-C does not work, that maybe we should set the default gnome-terminal configurable keybinding to something, giving people a chance to find it when they browse the menus.

Anyway, if someone wants to tell me where a new appVM qube gets its /home/user data to initialize the qube (I read that it does *not* copy the /home/user data from the template), I could take a shot at making a pull request for it.

**Comment by andrewdavidwong** [comment]
> > Sounds like a gnome-terminal bug.
> 
> 
> 
> If you mean bug in the general sense that it's specific to gnome-terminal configuration then yes.  In case this matters (and in case you don't already know this):  the "official" keybindings are "Shift-Ctrl-C".  This is what is in the default configuration on the "shortcuts" configuration screen.  the Ctrl-Insert keybinding seems to be a "secret" *second* keybinding that does not show up on the shortcuts configuration screen.  (i say "secret" cause it does not even appear on the gnome-terminal copy paste documentation at https://help.gnome.org/users/gnome-terminal/stable/txt-copy-paste.html )
> 
> 

Please look at the specific part of your message I was replying to.

> 
> Anyway, I was just proposing that if gnome-terminal is the default qubes terminal, and the default gnome-terminal keybinding of Shift-Ctrl-C does not work, that maybe we should set the default gnome-terminal configurable keybinding to something, giving people a chance to find it when they browse the menus.
> 

Yes, that's what this issue is for.

> 
> 
> Anyway, if someone wants to tell me where a new appVM qube gets its /home/user data to initialize the qube (I read that it does *not* copy the /home/user data from the template), I could take a shot at making a pull request for it.
> 
> 

/etc/skel

This is covered in the template documentation in further detail.

**Comment by ddevz** [comment]
Thanks!

After some research i found gnome does not use files and must be run as commands.  I did find the commands.  They are:

```
dconf write  /org/gnome/terminal/legacy/keybindings/copy "'<Primary>Insert'"
dconf write  /org/gnome/terminal/legacy/keybindings/paste  "'<Shift>Insert'"
```
However, if we are going to set the default, we might want to set it to something easier to use like:
```
dconf write  /org/gnome/terminal/legacy/keybindings/copy "'<Primary><Super>C'"
dconf write  /org/gnome/terminal/legacy/keybindings/paste  "'<Primary><Super>V'"
```
So I have 2 questions:

-1-  What should we set the default binding to?

-2-  Since it's a command and not just a file, is putting commands in .profile to set this up on first login a ok thing to do, or is that inappropriate?

**Comment by gw0** [comment]
In my opinion it would make most sense for Qubes to switch to "Windows-Key/Super" for its actions, because "Ctrl+Shift+c/v" is also used in other apps. Some key combinations with "Windows-Key/Super" that might come in handy:

- copy to secure clipboard "Super+c" (`/etc/qubes/guid.conf: secure_copy_sequence = "Mod4-c";`)
- paste to secure clipboard "Super+v" (`/etc/qubes/guid.conf: secure_paste_sequence = "Mod4-v";`)
- tile window to the left "Super+Left" (Settings Manager/Window Manager/Keyboard)
- tile window to the right "Super+Right" (Settings Manager/Window Manager/Keyboard)
- maximize window "Super+Up" (Settings Manager/Window Manager/Keyboard)
- shade window "Super+Down" (Settings Manager/Window Manager/Keyboard)

An alternative solution would also be to pass consecutive key presses of "Ctrl+Shift+c/v" down to the VM. This could be implemented by comparing the secret clipboard contents with what you want to copy (if same content, keys are passed to VM) or if there is nothing to paste (if clipboard empty, keys are passed to VM).

**Comment by andrewdavidwong** [comment]
> In my opinion it would make most sense for Qubes to switch to "Windows-Key/Super" for its actions, because "Ctrl+Shift+c/v" is also used in other apps.

https://github.com/QubesOS/qubes-issues/issues/4573#issuecomment-445100007

**Comment by gw0** [comment]
What about the idea that consecutive key presses within a short time period get passed down to the VM? The user experience would look like this:

- copy from terminal: Select text, press "Ctrl+Shift+c" twice within 500ms (first press would be an unnecessary secure copy, second press would be passed to the terminal to copy), press "Ctrl+Shift+c" again to perform secure copy
- paste into terminal: Press "Ctrl+Shift+v" twice within 500ms (first press would perform secure paste, second press would be passed to the terminal to paste)

After looking at the code (https://github.com/QubesOS/qubes-gui-daemon/blob/master/gui-daemon/xside.c#L991) this shouldn't be too difficult to implement. It would require extending `Ghandles` (adding a clipboard paste xevent timestamp and a flag to handle KeyRelease when pass-through gets triggered) and adding a configuration option for the pass-through interval into `guid.conf`.

**Comment by ddevz** [comment]
> > In my opinion it would make most sense for Qubes to switch to "Windows-Key/Super" for its actions, because "Ctrl+Shift+c/v" is also used in other apps.
> 
> [#4573 (comment)](https://github.com/QubesOS/qubes-issues/issues/4573#issuecomment-445100007)

If we assume that changing the default is not a option, then in my opinion it would make the most sense to create some kind of *standard* for a "alternate sequence".  By this I mean that people who want something other then shift-control-c still have to manually change it on their own in in /etc/qubes/guid.conf, but at least we all change it to the same thing!

This basically just becomes a change in documentation.

For a initial proposal for standardized alternate keys, I would propose @gw0 's idea of:

* copy to secure clipboard "Super+c" (/etc/qubes/guid.conf: secure_copy_sequence = "Mod4-c";)
* paste to secure clipboard "Super+v" (/etc/qubes/guid.conf: secure_paste_sequence = "Mod4-v";)

If this seems reasonable to others, I can change my personal settings to try these out and let everyone know if they are actually as convenient as they sound or if there are any issues with those particular keys.


(sorry for the delay in this response)

**Comment by ddevz** [comment]
I just tried to do a fork to create a proposed alternate in the comments of guid.conf, but found this:

# EDITING THIS FILE WILL NOT HAVE ANY EFFECT! It is not read by the GUI daemon
# anymore, and is left here for reference only. To configure the GUI daemon,
# use qvm-features (see 'man qvm-features' for details).

So It's not obvious to me where to document it anymore.

**Comment by andrewdavidwong** [comment]
> If we assume that changing the default is not a option, then in my opinion it would make the most sense to create some kind of _standard_ for a "alternate sequence". By this I mean that people who want something other then shift-control-c still have to manually change it on their own in in /etc/qubes/guid.conf, but at least we all change it to the same thing!
> 

Why?

**Comment by unman** [comment]
On Tue, Jul 27, 2021 at 11:29:54AM -0700, ddevz wrote:
> I just tried to do a fork to create a proposed alternate in the comments of guid.conf, but found this:
> 
> # EDITING THIS FILE WILL NOT HAVE ANY EFFECT! It is not read by the GUI daemon
> # anymore, and is left here for reference only. To configure the GUI daemon,
> # use qvm-features (see 'man qvm-features' for details).
> 
> So It's not obvious to me where to document it anymore.
> 

In the docs and in 'man qvm-features' for 4.1 , but I don't know what it
is you want to document.
I don't see particular value in proposing an alternative standard - as
was hashed over before, any proposed key combos will affect *someone*.
Leave it to users to choose one that works for them.

**Comment by ddevz** [comment]
This is basically fixed in 4.1rc1

In 4.1rc1, going to "global qubes settings", has a option to select between control-shift-v and control-Meta-v .  

Other then getting the message to people of where this setting is (I.E. documentation), I propose we can close this issue?

**Comment by andrewdavidwong** [comment]
> This is basically fixed in 4.1rc1
> 
> In 4.1rc1, going to "global qubes settings", has a option to select between control-shift-v and control-Meta-v .
> 
> Other then getting the message to people of where this setting is (I.E. documentation)

I think the answer is UX discoverability rather than documentation, but that's a separate matter either way.

> I propose we can close this issue?

Sounds good. Closing as resolved. If anyone believes this issue is not yet resolved, or if anyone is still affected by this issue, please leave a comment, and we'll be happy to reopen it. Thank you.
